In [0]:
import requests
import nltk
from bs4 import BeautifulSoup
nltk.download('punkt')
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
import random


In [0]:
URL = 'https://www.thestar.com/news/gta/2019/12/31/with-291-people-shot-2019-is-closing-as-torontos-bloodiest-year-on-record-for-overall-gun-violence.html'

In [0]:
page = requests.get(URL)
soup = BeautifulSoup(page.content, 'html.parser')
soup.title.text

'With 292 people shot, 2019 comes to a close as Toronto’s bloodiest year on record for overall gun violence | The Star'

Code below works for Toronto Star articles. Try to modify for Toronto Sun too.

In [0]:
results = soup.find(class_='c-article-body__content') 
#print(results.prettify())
results_text = [tag.get_text().strip() for tag in results]
sentence_list = [sentence for sentence in results_text if not '\n' in sentence]
sentence_list = [sentence for sentence in sentence_list if '.' in sentence]
article = ' '.join(sentence_list)
print(article)

Toronto’s bloodiest year for gun violence is coming to a close. It started early on Jan. 1, 2019, when a man was shot at Oakwood Avenue and Vaughan Road. He was rushed to a trauma centre and survived. And the violence continued until the last day of the year, with a double shooting near Jane Street and Finch Avenue West in the early hours of New Year’s Eve, and another man shot and killed in the evening. The two victims of the first shooting, both men, were rushed to hospital in serious condition. The victim of the second died in hospital after being shot in Scarborough at Neilson Road and Crow Trail, near Finch Avenue East. As of 9 p.m. on New Year’s Eve, 292 people had been  killed or injured in shootings in the city, 56 more than the previous year, which itself set a record for the most people shot in police data that goes back 15 years. Despite a more than 20 per cent increase in the number of shootings compared to 2018, this year saw seven fewer people killed by guns. Experts say 

In [0]:
tokens = [t for t in article.split()]
print(tokens)

['Toronto’s', 'bloodiest', 'year', 'for', 'gun', 'violence', 'is', 'coming', 'to', 'a', 'close.', 'It', 'started', 'early', 'on', 'Jan.', '1,', '2019,', 'when', 'a', 'man', 'was', 'shot', 'at', 'Oakwood', 'Avenue', 'and', 'Vaughan', 'Road.', 'He', 'was', 'rushed', 'to', 'a', 'trauma', 'centre', 'and', 'survived.', 'And', 'the', 'violence', 'continued', 'until', 'the', 'last', 'day', 'of', 'the', 'year,', 'with', 'a', 'double', 'shooting', 'near', 'Jane', 'Street', 'and', 'Finch', 'Avenue', 'West', 'in', 'the', 'early', 'hours', 'of', 'New', 'Year’s', 'Eve,', 'and', 'another', 'man', 'shot', 'and', 'killed', 'in', 'the', 'evening.', 'The', 'two', 'victims', 'of', 'the', 'first', 'shooting,', 'both', 'men,', 'were', 'rushed', 'to', 'hospital', 'in', 'serious', 'condition.', 'The', 'victim', 'of', 'the', 'second', 'died', 'in', 'hospital', 'after', 'being', 'shot', 'in', 'Scarborough', 'at', 'Neilson', 'Road', 'and', 'Crow', 'Trail,', 'near', 'Finch', 'Avenue', 'East.', 'As', 'of', '9', '

In [0]:
from nltk.tokenize import sent_tokenize, word_tokenize
print(word_tokenize(article))

['Toronto', '’', 's', 'bloodiest', 'year', 'for', 'gun', 'violence', 'is', 'coming', 'to', 'a', 'close', '.', 'It', 'started', 'early', 'on', 'Jan.', '1', ',', '2019', ',', 'when', 'a', 'man', 'was', 'shot', 'at', 'Oakwood', 'Avenue', 'and', 'Vaughan', 'Road', '.', 'He', 'was', 'rushed', 'to', 'a', 'trauma', 'centre', 'and', 'survived', '.', 'And', 'the', 'violence', 'continued', 'until', 'the', 'last', 'day', 'of', 'the', 'year', ',', 'with', 'a', 'double', 'shooting', 'near', 'Jane', 'Street', 'and', 'Finch', 'Avenue', 'West', 'in', 'the', 'early', 'hours', 'of', 'New', 'Year', '’', 's', 'Eve', ',', 'and', 'another', 'man', 'shot', 'and', 'killed', 'in', 'the', 'evening', '.', 'The', 'two', 'victims', 'of', 'the', 'first', 'shooting', ',', 'both', 'men', ',', 'were', 'rushed', 'to', 'hospital', 'in', 'serious', 'condition', '.', 'The', 'victim', 'of', 'the', 'second', 'died', 'in', 'hospital', 'after', 'being', 'shot', 'in', 'Scarborough', 'at', 'Neilson', 'Road', 'and', 'Crow', 'Tra

In [0]:
word_tokens = word_tokenize(article)
stop_words = set(stopwords.words('english'))
filtered_article = [w for w in word_tokens if not w in stop_words]
filtered_sentence = []

for w in word_tokens:
  if w not in stop_words:
    filtered_sentence.append(w)

print(filtered_sentence)

['Toronto', '’', 'bloodiest', 'year', 'gun', 'violence', 'coming', 'close', '.', 'It', 'started', 'early', 'Jan.', '1', ',', '2019', ',', 'man', 'shot', 'Oakwood', 'Avenue', 'Vaughan', 'Road', '.', 'He', 'rushed', 'trauma', 'centre', 'survived', '.', 'And', 'violence', 'continued', 'last', 'day', 'year', ',', 'double', 'shooting', 'near', 'Jane', 'Street', 'Finch', 'Avenue', 'West', 'early', 'hours', 'New', 'Year', '’', 'Eve', ',', 'another', 'man', 'shot', 'killed', 'evening', '.', 'The', 'two', 'victims', 'first', 'shooting', ',', 'men', ',', 'rushed', 'hospital', 'serious', 'condition', '.', 'The', 'victim', 'second', 'died', 'hospital', 'shot', 'Scarborough', 'Neilson', 'Road', 'Crow', 'Trail', ',', 'near', 'Finch', 'Avenue', 'East', '.', 'As', '9', 'p.m.', 'New', 'Year', '’', 'Eve', ',', '292', 'people', 'killed', 'injured', 'shootings', 'city', ',', '56', 'previous', 'year', ',', 'set', 'record', 'people', 'shot', 'police', 'data', 'goes', 'back', '15', 'years', '.', 'Despite', '

In [0]:
clean_tokens = tokens[:]

for token in tokens:
  if token in stopwords.words('english'):
    
    clean_tokens.remove(token)

freq = nltk.FreqDist(clean_tokens)
for key,val in freq.items():
  print(str(key) + ':' + str(val))

freq.plot(20, cumulative=False)

Need to determine how to lemmatize my data set. The below code doesn't work with a list. 

In [0]:
lemmatizer = WordNetLemmatizer()

print(lemmatizer.lemmatize(filtered_sentence))

Building the dataframe: tstar_articles

In [0]:
import pandas as pd
url_file = 'https://github.com/MarissaFosse/ryersoncapstone/raw/master/DailyNewsArticles.xlsx'

tstar_articles = pd.read_excel(url_file, "TorontoStar Articles", header=0) 
#print (tstar_articles)

tstar_articles.describe()

Date  ...                                                URL
count                   147  ...                                                147
unique                   51  ...                                                147
top     2020-01-22 00:00:00  ...  https://www.thestar.com/news/gta/2019/12/31/si...
freq                      7  ...                                                  1
first   2020-01-01 00:00:00  ...                                                NaN
last    2020-02-29 00:00:00  ...                                                NaN

[6 rows x 5 columns]

In [0]:
tstar_articles.head()

Date  ...                                                URL
0 2020-01-01  ...  https://www.thestar.com/news/gta/2019/12/31/wi...
1 2020-01-01  ...  https://www.thestar.com/news/crime/2019/12/31/...
2 2020-01-01  ...  https://www.thestar.com/news/gta/2019/12/31/si...
3 2020-01-04  ...  https://www.thestar.com/news/crime/2020/01/02/...
4 2020-01-04  ...  https://www.thestar.com/news/gta/2020/01/02/si...

[5 rows x 5 columns]

**Using the dataframe above, how can I write a Python script to open the URL, scrape the text, append the article text to the row as another column (or perhaps a linked text file?) and then move to the next row until all 147 have been scraped?**

In [0]:
nltk.download('movie_reviews')
from nltk.corpus import movie_reviews

documents = [(list(movie_reviews.words(fileid)), category)
             for category in movie_reviews.categories()
             for fileid in movie_reviews.fileids(category)]

#documents = []
#for category in movie_reviews.categories():
#   for fileid in movie_reviews.fileids(category):
#     documents.append(movie_reviews.words(fileid), category)


random.shuffle(documents)

#print(documents[1])

all_words = []
for w in movie_reviews.words():
    all_words.append(w.lower())

all_words = nltk.FreqDist(all_words)

word_features = list(all_words.keys())[:3000]

def find_features(document):
  words = set(document) #a set is one iteration of all words
  features = {}
  for w in word_features:
    features[w] = (w in words)
  return features

print((find_features(movie_reviews.words('neg/cv000_29416.txt'))))

featuresets = [(find_features(rev), category) for (rev, category) in documents]

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!
{'plot': True, ':': True, 'two': True, 'teen': True, 'couples': True, 'go': True, 'to': True, 'a': True, 'church': True, 'party': True, ',': True, 'drink': True, 'and': True, 'then': True, 'drive': True, '.': True, 'they': True, 'get': True, 'into': True, 'an': True, 'accident': True, 'one': True, 'of': True, 'the': True, 'guys': True, 'dies': True, 'but': True, 'his': True, 'girlfriend': True, 'continues': True, 'see': True, 'him': True, 'in': True, 'her': True, 'life': True, 'has': True, 'nightmares': True, 'what': True, "'": True, 's': True, 'deal': True, '?': True, 'watch': True, 'movie': True, '"': True, 'sorta': True, 'find': True, 'out': True, 'critique': True, 'mind': True, '-': True, 'fuck': True, 'for': True, 'generation': True, 'that': True, 'touches': True, 'on': True, 'very': True, 'cool': True, 'idea': True, 'presents': True, 'it': True, 'bad': Tr

In [0]:
training_set = featuresets[:1900] #up to but not including 1900'th review
testing_set = featuresets[1900:]

classifier = nltk.NaiveBayesClassifier.train(training_set)
print("Classifier accuracy percent:",(nltk.classify.accuracy(classifier, testing_set))*100)